src/chat.rs

```diff
@@ -1291,6 +1295,75 @@ impl ChatId {
             .unwrap_or_default();
         Ok(protection_status)
     }
+
+    /// Returns the sort timestamp for a protection message.
+    ///
+    /// `timestamp` should be normally the "sent" timestamp of the message caused the chat
+    /// protection state change.
+    pub(crate) async fn calc_protection_msg_sort_timestamp(
+        self,
+        context: &Context,
+        timestamp: i64,
+    ) -> Result<i64> {
+        let sort_to_bottom = true;
+        self.calc_sort_timestamp(context, timestamp, sort_to_bottom, false)
+            .await
+    }
+
+    /// Returns the sort timestamp for a new message in the chat.
+    ///
+    /// `message_timestamp` should be either the message "sent" timestamp or a timestamp of the
+    /// corresponding event in case of a system message (usually the current system time).
+    /// `always_sort_to_bottom` makes this ajust the returned timestamp up so that the message goes
+    /// to the chat bottom.
+    /// `incoming` -- whether the message is incoming.
+    pub(crate) async fn calc_sort_timestamp(
+        self,
+        context: &Context,
+        message_timestamp: i64,
+        always_sort_to_bottom: bool,
+        incoming: bool,
+    ) -> Result<i64> {
+        let mut sort_timestamp = cmp::min(message_timestamp, smeared_time(context));
+
+        let last_msg_time: Option<i64> = if always_sort_to_bottom {
+            // get newest message for this chat
+            context
+                .sql
+                .query_get_value(
+                    "SELECT MAX(timestamp) FROM msgs WHERE chat_id=? AND state!=?",
+                    (self, MessageState::OutDraft),
+                )
+                .await?
+        } else if incoming {
+            // get newest non fresh message for this chat.
+
+            // If a user hasn't been online for some time, the Inbox is fetched first and then the
+            // Sentbox. In order for Inbox and Sent messages to be allowed to mingle, outgoing
+            // messages are purely sorted by their sent timestamp. NB: The Inbox must be fetched
+            // first otherwise Inbox messages would be always below old Sentbox messages. We could
+            // take in the query below only incoming messages, but then new incoming messages would
+            // mingle with just sent outgoing ones and apear somewhere in the middle of the chat.
+            context
+                .sql
+                .query_get_value(
+                    "SELECT MAX(timestamp) FROM msgs \
+                    WHERE chat_id=? AND hidden=0 AND state>? AND state!=?",
+                    (self, MessageState::InFresh, MessageState::OutDraft),
```

Comment:
Why check whether it's a draft after we already checked whether it's hidden? Drafts are set as `hidden`, otherwise they would be shown in the chat, just as location messages, reactions, etc.